include/jsoncons/json.hpp

```diff
@@ -654,21 +658,88 @@ class basic_json
         };
 
     private:
-        static const size_t data_size = static_max<sizeof(uinteger_data),sizeof(double_data),sizeof(small_string_data), sizeof(string_data), sizeof(array_data), sizeof(object_data)>::value;
-        static const size_t data_align = static_max<JSONCONS_ALIGNOF(uinteger_data),JSONCONS_ALIGNOF(double_data),JSONCONS_ALIGNOF(small_string_data),JSONCONS_ALIGNOF(string_data),JSONCONS_ALIGNOF(array_data),JSONCONS_ALIGNOF(object_data)>::value;
 
-        typedef typename std::aligned_storage<data_size,data_align>::type data_t;
+        union
+        {
+            base_data base_data_;
+            null_data null_data_;
+            bool_data bool_data_;
+            integer_data integer_data_;
+            uinteger_data uinteger_data_;
+            double_data double_data_;
+            small_string_data small_string_data_;
+            string_data string_data_;
+            byte_string_data byte_string_data_;
+            array_data array_data_;
+            object_data object_data_;
+            empty_object_data empty_object_data_;
+        };
+
+        template <class VariantType, class... Args>
+        void construct_var(Args&&... args)
+        {
+            new (&as<VariantType>()) VariantType(std::forward<Args>(args)...);
+        }
+
+        template <class VariantType>
+        void destroy_var()
+        {
+            as<VariantType>().~VariantType();
+        }
 
-        data_t data_;
     public:
+        template <class T> T&          as();
+        template <> base_data&         as<base_data>()         { return base_data_; }
+        template <> null_data&         as<null_data>()         { return null_data_; }
+        template <> bool_data&         as<bool_data>()         { return bool_data_; }
+        template <> integer_data&      as<integer_data>()      { return integer_data_; }
+        template <> uinteger_data&     as<uinteger_data>()     { return uinteger_data_; }
+        template <> double_data&       as<double_data>()       { return double_data_; }
+        template <> small_string_data& as<small_string_data>() { return small_string_data_; }
+        template <> string_data&       as<string_data>()       { return string_data_; }
+        template <> byte_string_data&  as<byte_string_data>()  { return byte_string_data_; }
+        template <> array_data&        as<array_data>()        { return array_data_; }
+        template <> object_data&       as<object_data>()       { return object_data_; }
+        template <> empty_object_data& as<empty_object_data>() { return empty_object_data_; }
+
+        template <class T> const T& as_c() const;
```

Comment:
This is a bit odd. This template function should be simply named `as() const`, which is the counter part to the above non const template member function. But for some weird reason this would completely break IntelliSense in VisualStudio 2017, meaning no more syntax highlighting / auto completion / etc in the entire project.

So I named it `as_c() const` for now.